File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/auth/NativeAuthCoordinator.kt
package expo.modules.vescapecore.auth

import android.content.Context
import expo.modules.vescapecore.appstatus.AppStatusCoordinator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * Reusable native authenticated HTTP boundary.
 * @parity /modules/vescape-core/ios/auth/NativeAuthCoordinator.swift
 */
class NativeAuthCoordinator(private val context: Context) {
  private val store = DeviceCredentialStore(context)
  private val client = OkHttpClient.Builder().callTimeout(10, TimeUnit.SECONDS).build()

  fun stateMap(): Map<String, Any?> {
    val credential = store.read()
    return mapOf(
      "state" to store.state().slug,
      "accountId" to credential?.accountId,
      "expiresAt" to credential?.expiresAt,
    )
  }

  suspend fun provision(
    serverUrl: String,
    token: String,
    accountId: String,
  ): Map<String, Any?> = withContext(Dispatchers.IO) {
    val candidate = DeviceCredential(serverUrl.trimEnd('/'), token, accountId, null)
    val request = authenticatedRequest(candidate, "/api/account").get().build()
    client.newCall(request).execute().use { response ->
      if (response.code == 401) {
        store.reject()
        throw IllegalStateException("Device credential rejected")
      }
      check(response.isSuccessful) { "Account verification failed (${response.code})" }
      val returnedId = response.body?.string()
        ?.let { org.json.JSONObject(it).optString("id") }
        .orEmpty()
      check(returnedId == accountId) { "Account verification mismatch" }
      store.write(candidate)
    }
    AppStatusCoordinator.get(context).refresh()
    stateMap()
  }

  suspend fun revoke() = withContext(Dispatchers.IO) {
    val credential = store.read() ?: return@withContext
    val request = authenticatedRequest(credential, "/api/auth/device-tokens/current")
      .delete()
      .build()
    client.newCall(request).execute().use { response ->
      check(response.isSuccessful || response.code == 401) {
        "Device credential revocation failed (${response.code})"
      }
    }
    store.clear()
  }

  fun clear() = store.clear()

  private fun authenticatedRequest(
    credential: DeviceCredential,
    path: String,
  ): Request.Builder = Request.Builder()
    .url("${credential.serverUrl}$path")
    .header("Authorization", "Bearer ${credential.token}")
    .header(AppStatusCoordinator.APP_VERSION_HEADER, AppStatusCoordinator.get(context).appVersion)

  companion object {
    @Volatile private var instance: NativeAuthCoordinator? = null
    fun get(context: Context): NativeAuthCoordinator =
      instance ?: synchronized(this) {
        instance ?: NativeAuthCoordinator(context.applicationContext).also { instance = it }
      }
  }
}

File: modules/vescape-core/ios/auth/NativeAuthCoordinator.swift
import Foundation

/// Reusable native authenticated HTTP boundary.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/auth/NativeAuthCoordinator.kt
final class NativeAuthCoordinator {
  static let shared = NativeAuthCoordinator()
  private let store = DeviceCredentialStore.shared

  func stateMap() -> [String: Any?] {
    let credential = store.read()
    return [
      "state": store.state().rawValue,
      "accountId": credential?.accountId,
      "expiresAt": credential?.expiresAt,
    ]
  }

  func provision(
    serverUrl: String,
    token: String,
    accountId: String
  ) async throws -> [String: Any?] {
    let credential = DeviceCredential(
      serverUrl: serverUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")),
      token: token,
      accountId: accountId,
      expiresAt: nil
    )
    let (data, response) = try await request(credential, path: "/api/account")
    if response.statusCode == 401 {
      store.reject()
      throw NSError(domain: "NativeAuth", code: 401)
    }
    guard (200..<300).contains(response.statusCode),
          let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          json["id"] as? String == accountId
    else { throw NSError(domain: "NativeAuth", code: response.statusCode) }
    try store.write(credential)
    await MainActor.run {
      AppStatusCoordinator.shared.refresh()
    }
    return stateMap()
  }

  func revoke() async throws {
    guard let credential = store.read() else { return }
    let (_, response) = try await request(
      credential,
      path: "/api/auth/device-tokens/current",
      method: "DELETE"
    )
    guard (200..<300).contains(response.statusCode) || response.statusCode == 401 else {
      throw NSError(domain: "NativeAuth", code: response.statusCode)
    }
    store.clear()
  }

  func clear() { store.clear() }

  private func request(
    _ credential: DeviceCredential,
    path: String,
    method: String = "GET"
  ) async throws -> (Data, HTTPURLResponse) {
    guard let url = URL(string: "\(credential.serverUrl)\(path)") else {
      throw NSError(domain: "NativeAuth", code: -1)
    }
    var request = URLRequest(url: url)
    request.httpMethod = method
    request.timeoutInterval = 10
    request.setValue("Bearer \(credential.token)", forHTTPHeaderField: "Authorization")
    request.setValue(
      AppStatusCoordinator.installedMarketingVersion(),
      forHTTPHeaderField: AppStatusCoordinator.appVersionHeader
    )
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else {
      throw NSError(domain: "NativeAuth", code: -2)
    }
    return (data, http)
  }
}
